google/cloud/storage/_helpers.py

```diff
@@ -312,3 +402,44 @@ def _convert_to_timestamp(value):
     utc_naive = value.replace(tzinfo=None) - value.utcoffset()
     mtime = (utc_naive - datetime(1970, 1, 1)).total_seconds()
     return mtime
+
+
+def _add_generation_match_parameters(parameters, **match_parameters):
+    """Add generation match parameters into the given parameters list.
+
+    :type parameters: list or dict
+    :param parameters: Parameters list or dict.
+
+    :type match_parameters: dict
+    :param match_parameters: if*generation*match parameters to add.
+    """
+    for snakecase_name, camelcase_name in _GENERATION_MATCH_PARAMETERS:
+        value = match_parameters.get(snakecase_name)
+
+        if value is not None:
+            if isinstance(parameters, list):
+                parameters.append((camelcase_name, value))
+
+            elif isinstance(parameters, dict):
+                parameters[camelcase_name] = value
```

Comment:
Added one more condition into helper, as there are methods in which query_params is dict, not list with (key, value) pairs.

Reply:
May I suggest we use duck typing here and just try to append, catch a failure, and then use the dict access, in order to safely handle any list-alikes or dict-alikes that may sneak into the code later? Also, might be good to error out if neither the list or dict access works.

Reply:
Just double checking, but @andrewsg are you okay with how this was resolved?

Reply:
Yes, adding the "else" clause works for me.